# Project 2: Data Import - Working with Web APIs and JSON (Movies Dataset)

# Project Brief for Self-Coders

Here you´ll have the opportunity to code major parts of Project 2 on your own. If you need any help or inspiration, have a look at the Videos or the Jupyter Notebook with the full code. <br> <br>
Keep in mind that it´s all about __getting the right results/conclusions__. It´s not about finding the identical code. Things can be coded in many different ways. Even if you come to the same conclusions, it´s very unlikely that we have the very same code. 

## Importing Data from JSON files 

1. __Import__ the json files __blockbusters.json__, __blockbusters2.json__, __blockbusters3.json__ and load the datasets into Pandas DataFrames.


In [3]:
import pandas as pd
import json

In [4]:
with open("blockbusters.json") as f:
    data = json.load(f)

In [5]:
data

[{'title': 'Avengers: Endgame',
  'id': 299534,
  'revenue': 2797800564,
  'genres': [{'id': 12, 'name': 'Adventure'},
   {'id': 878, 'name': 'Science Fiction'},
   {'id': 28, 'name': 'Action'}],
  'belongs_to_collection': {'id': 86311,
   'name': 'The Avengers Collection',
   'poster_path': '/yFSIUVTCvgYrpalUktulvk3Gi5Y.jpg',
   'backdrop_path': '/zuW6fOiusv4X9nnW3paHGfXcSll.jpg'},
  'runtime': 181},
 {'title': 'Avatar',
  'id': 19995,
  'revenue': 2787965087,
  'genres': [{'id': 28, 'name': 'Action'},
   {'id': 12, 'name': 'Adventure'},
   {'id': 14, 'name': 'Fantasy'},
   {'id': 878, 'name': 'Science Fiction'}],
  'belongs_to_collection': {'id': 87096,
   'name': 'Avatar Collection',
   'poster_path': '/nslJVsO58Etqkk17oXMuVK4gNOF.jpg',
   'backdrop_path': '/8nCr9W7sKus2q9PLbYsnT7iCkuT.jpg'},
  'runtime': 162},
 {'title': 'Star Wars: The Force Awakens',
  'id': 140607,
  'revenue': 2068223624,
  'genres': [{'id': 28, 'name': 'Action'},
   {'id': 12, 'name': 'Adventure'},
   {'id': 8

In [6]:
type(data)

list

In [7]:
len(data)

18

In [8]:
data[0]

{'title': 'Avengers: Endgame',
 'id': 299534,
 'revenue': 2797800564,
 'genres': [{'id': 12, 'name': 'Adventure'},
  {'id': 878, 'name': 'Science Fiction'},
  {'id': 28, 'name': 'Action'}],
 'belongs_to_collection': {'id': 86311,
  'name': 'The Avengers Collection',
  'poster_path': '/yFSIUVTCvgYrpalUktulvk3Gi5Y.jpg',
  'backdrop_path': '/zuW6fOiusv4X9nnW3paHGfXcSll.jpg'},
 'runtime': 181}

In [9]:
df = pd.DataFrame(data)
df

title      id     revenue  \
0                              Avengers: Endgame  299534  2797800564   
1                                         Avatar   19995  2787965087   
2                   Star Wars: The Force Awakens  140607  2068223624   
3                         Avengers: Infinity War  299536  2046239637   
4                                        Titanic     597  1845034188   
5                                 Jurassic World  135397  1671713208   
6                                  The Lion King  420818  1656943394   
7                                   The Avengers   24428  1519557910   
8                                      Furious 7  168259  1506249360   
9                        Avengers: Age of Ultron   99861  1405403694   
10                                 Black Panther  284054  1346739107   
11  Harry Potter and the Deathly Hallows: Part 2   12445  1341511219   
12                      Star Wars: The Last Jedi  181808  1332539889   
13                                     Frozen II  330457  1330764959   
14                Jurassic World: Fallen Kingdom  351286  1303459585   
15                                        Frozen  109445  1274219009   
16                          Beauty and the Beast  321612  1263521126   
17                                 Incredibles 2  260513  1241891456   

                                               genres  \
0   [{'id': 12, 'name': 'Adventure'}, {'id': 878, ...   
1   [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
2   [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
3   [{'id': 12, 'name': 'Adventure'}, {'id': 28, '...   
4   [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...   
5   [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
6   [{'id': 12, 'name': 'Adventure'}, {'id': 10751...   
7   [{'id': 878, 'name': 'Science Fiction'}, {'id'...   
8   [{'id': 28, 'name': 'Action'}, {'id': 53, 'nam...   
9   [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
10  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
11  [{'id': 14, 'name': 'Fantasy'}, {'id': 12, 'na...   
12  [{'id': 14, 'name': 'Fantasy'}, {'id': 12, 'na...   
13  [{'id': 16, 'name': 'Animation'}, {'id': 10751...   
14  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
15  [{'id': 16, 'name': 'Animation'}, {'id': 12, '...   
16  [{'id': 10751, 'name': 'Family'}, {'id': 14, '...   
17  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   

                                belongs_to_collection  runtime  
0   {'id': 86311, 'name': 'The Avengers Collection...      181  
1   {'id': 87096, 'name': 'Avatar Collection', 'po...      162  
2   {'id': 10, 'name': 'Star Wars Collection', 'po...      136  
3   {'id': 86311, 'name': 'The Avengers Collection...      149  
4                                                None      194  
5   {'id': 328, 'name': 'Jurassic Park Collection'...      124  
6                                                None      118  
7   {'id': 86311, 'name': 'The Avengers Collection...      143  
8   {'id': 9485, 'name': 'The Fast and the Furious...      137  
9   {'id': 86311, 'name': 'The Avengers Collection...      141  
10  {'id': 529892, 'name': 'Black Panther Collecti...      134  
11  {'id': 1241, 'name': 'Harry Potter Collection'...      130  
12  {'id': 10, 'name': 'Star Wars Collection', 'po...      152  
13  {'id': 386382, 'name': 'Frozen Collection', 'p...      104  
14  {'id': 328, 'name': 'Jurassic Park Collection'...      129  
15  {'id': 386382, 'name': 'Frozen Collection', 'p...      102  
16                                               None      129  
17  {'id': 468222, 'name': 'The Incredibles Collec...      118

In [10]:
df = pd.read_json("blockbusters.json")
df

title      id     revenue  \
0                              Avengers: Endgame  299534  2797800564   
1                                         Avatar   19995  2787965087   
2                   Star Wars: The Force Awakens  140607  2068223624   
3                         Avengers: Infinity War  299536  2046239637   
4                                        Titanic     597  1845034188   
5                                 Jurassic World  135397  1671713208   
6                                  The Lion King  420818  1656943394   
7                                   The Avengers   24428  1519557910   
8                                      Furious 7  168259  1506249360   
9                        Avengers: Age of Ultron   99861  1405403694   
10                                 Black Panther  284054  1346739107   
11  Harry Potter and the Deathly Hallows: Part 2   12445  1341511219   
12                      Star Wars: The Last Jedi  181808  1332539889   
13                                     Frozen II  330457  1330764959   
14                Jurassic World: Fallen Kingdom  351286  1303459585   
15                                        Frozen  109445  1274219009   
16                          Beauty and the Beast  321612  1263521126   
17                                 Incredibles 2  260513  1241891456   

                                               genres  \
0   [{'id': 12, 'name': 'Adventure'}, {'id': 878, ...   
1   [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
2   [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
3   [{'id': 12, 'name': 'Adventure'}, {'id': 28, '...   
4   [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...   
5   [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
6   [{'id': 12, 'name': 'Adventure'}, {'id': 10751...   
7   [{'id': 878, 'name': 'Science Fiction'}, {'id'...   
8   [{'id': 28, 'name': 'Action'}, {'id': 53, 'nam...   
9   [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
10  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
11  [{'id': 14, 'name': 'Fantasy'}, {'id': 12, 'na...   
12  [{'id': 14, 'name': 'Fantasy'}, {'id': 12, 'na...   
13  [{'id': 16, 'name': 'Animation'}, {'id': 10751...   
14  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
15  [{'id': 16, 'name': 'Animation'}, {'id': 12, '...   
16  [{'id': 10751, 'name': 'Family'}, {'id': 14, '...   
17  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   

                                belongs_to_collection  runtime  
0   {'id': 86311, 'name': 'The Avengers Collection...      181  
1   {'id': 87096, 'name': 'Avatar Collection', 'po...      162  
2   {'id': 10, 'name': 'Star Wars Collection', 'po...      136  
3   {'id': 86311, 'name': 'The Avengers Collection...      149  
4                                                None      194  
5   {'id': 328, 'name': 'Jurassic Park Collection'...      124  
6                                                None      118  
7   {'id': 86311, 'name': 'The Avengers Collection...      143  
8   {'id': 9485, 'name': 'The Fast and the Furious...      137  
9   {'id': 86311, 'name': 'The Avengers Collection...      141  
10  {'id': 529892, 'name': 'Black Panther Collecti...      134  
11  {'id': 1241, 'name': 'Harry Potter Collection'...      130  
12  {'id': 10, 'name': 'Star Wars Collection', 'po...      152  
13  {'id': 386382, 'name': 'Frozen Collection', 'p...      104  
14  {'id': 328, 'name': 'Jurassic Park Collection'...      129  
15  {'id': 386382, 'name': 'Frozen Collection', 'p...      102  
16                                               None      129  
17  {'id': 468222, 'name': 'The Incredibles Collec...      118

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18 entries, 0 to 17
Data columns (total 6 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   title                  18 non-null     object
 1   id                     18 non-null     int64 
 2   revenue                18 non-null     int64 
 3   genres                 18 non-null     object
 4   belongs_to_collection  15 non-null     object
 5   runtime                18 non-null     int64 
dtypes: int64(3), object(3)
memory usage: 992.0+ bytes


In [12]:
df.genres

0     [{'id': 12, 'name': 'Adventure'}, {'id': 878, ...
1     [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...
2     [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...
3     [{'id': 12, 'name': 'Adventure'}, {'id': 28, '...
4     [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...
5     [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...
6     [{'id': 12, 'name': 'Adventure'}, {'id': 10751...
7     [{'id': 878, 'name': 'Science Fiction'}, {'id'...
8     [{'id': 28, 'name': 'Action'}, {'id': 53, 'nam...
9     [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...
10    [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...
11    [{'id': 14, 'name': 'Fantasy'}, {'id': 12, 'na...
12    [{'id': 14, 'name': 'Fantasy'}, {'id': 12, 'na...
13    [{'id': 16, 'name': 'Animation'}, {'id': 10751...
14    [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...
15    [{'id': 16, 'name': 'Animation'}, {'id': 12, '...
16    [{'id': 10751, 'name': 'Family'}, {'id': 14, '...
17    [{'id': 28, 'name': 'Action'}, {'id': 12, 

In [13]:
df.genres[0]

[{'id': 12, 'name': 'Adventure'},
 {'id': 878, 'name': 'Science Fiction'},
 {'id': 28, 'name': 'Action'}]

In [14]:
df.belongs_to_collection[0]

{'id': 86311,
 'name': 'The Avengers Collection',
 'poster_path': '/yFSIUVTCvgYrpalUktulvk3Gi5Y.jpg',
 'backdrop_path': '/zuW6fOiusv4X9nnW3paHGfXcSll.jpg'}

In [15]:
pd.json_normalize(data = data, sep = "_").genres[0]

[{'id': 12, 'name': 'Adventure'},
 {'id': 878, 'name': 'Science Fiction'},
 {'id': 28, 'name': 'Action'}]

In [16]:
pd.json_normalize(data = data, record_path = "genres", meta =["title", "id"], record_prefix = "genre_")

genre_id       genre_name                                         title  \
0         12        Adventure                             Avengers: Endgame   
1        878  Science Fiction                             Avengers: Endgame   
2         28           Action                             Avengers: Endgame   
3         28           Action                                        Avatar   
4         12        Adventure                                        Avatar   
5         14          Fantasy                                        Avatar   
6        878  Science Fiction                                        Avatar   
7         28           Action                  Star Wars: The Force Awakens   
8         12        Adventure                  Star Wars: The Force Awakens   
9        878  Science Fiction                  Star Wars: The Force Awakens   
10        14          Fantasy                  Star Wars: The Force Awakens   
11        12        Adventure                        Avengers: Infinity War   
12        28           Action                        Avengers: Infinity War   
13       878  Science Fiction                        Avengers: Infinity War   
14        18            Drama                                       Titanic   
15     10749          Romance                                       Titanic   
16        53         Thriller                                       Titanic   
17        28           Action                                Jurassic World   
18        12        Adventure                                Jurassic World   
19       878  Science Fiction                                Jurassic World   
20        53         Thriller                                Jurassic World   
21        12        Adventure                                 The Lion King   
22     10751           Family                                 The Lion King   
23       878  Science Fiction                                  The Avengers   
24        28           Action                                  The Avengers   
25        12        Adventure                                  The Avengers   
26        28           Action                                     Furious 7   
27        53         Thriller                                     Furious 7   
28        28           Action                       Avengers: Age of Ultron   
29        12        Adventure                       Avengers: Age of Ultron   
30       878  Science Fiction                       Avengers: Age of Ultron   
31        28           Action                                 Black Panther   
32        12        Adventure                                 Black Panther   
33        14          Fantasy                                 Black Panther   
34       878  Science Fiction                                 Black Panther   
35        14          Fantasy  Harry Potter and the Deathly Hallows: Part 2   
36        12        Adventure  Harry Potter and the Deathly Hallows: Part 2   
37        14          Fantasy                      Star Wars: The Last Jedi   
38        12        Adventure                      Star Wars: The Last Jedi   
39       878  Science Fiction                      Star Wars: The Last Jedi   
40        28           Action                      Star Wars: The Last Jedi   
41        16        Animation                                     Frozen II   
42     10751           Family                                     Frozen II   
43        12        Adventure                                     Frozen II   
44        28           Action                Jurassic World: Fallen Kingdom   
45        12        Adventure                Jurassic World: Fallen Kingdom   
46       878  Science Fiction                Jurassic World: Fallen Kingdom   
47        16        Animation                                        Frozen   
48        12        Adventure                                        Frozen   
49     10751           Family                         

In [17]:
with open("blockbusters3.json") as f:
    data = json.load(f)

In [18]:
data

{'columns': ['title',
  'id',
  'revenue',
  'genres',
  'belongs_to_collection',
  'runtime'],
 'index': [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17],
 'data': [['Avengers: Endgame',
   299534,
   2797800564,
   [{'id': 12, 'name': 'Adventure'},
    {'id': 878, 'name': 'Science Fiction'},
    {'id': 28, 'name': 'Action'}],
   {'id': 86311,
    'name': 'The Avengers Collection',
    'poster_path': '/yFSIUVTCvgYrpalUktulvk3Gi5Y.jpg',
    'backdrop_path': '/zuW6fOiusv4X9nnW3paHGfXcSll.jpg'},
   181],
  ['Avatar',
   19995,
   2787965087,
   [{'id': 28, 'name': 'Action'},
    {'id': 12, 'name': 'Adventure'},
    {'id': 14, 'name': 'Fantasy'},
    {'id': 878, 'name': 'Science Fiction'}],
   {'id': 87096,
    'name': 'Avatar Collection',
    'poster_path': '/nslJVsO58Etqkk17oXMuVK4gNOF.jpg',
    'backdrop_path': '/8nCr9W7sKus2q9PLbYsnT7iCkuT.jpg'},
   162],
  ['Star Wars: The Force Awakens',
   140607,
   2068223624,
   [{'id': 28, 'name': 'Action'},
    {'id': 12, 'name':

In [19]:
type(data)

dict

In [20]:
len(data)

3

In [21]:
data[0]

KeyError: 0

In [ ]:
df = pd.DataFrame(data)
df

ValueError: All arrays must be of the same length

In [ ]:
df = pd.read_json('blockbusters3.json', orient='split')
df

title      id     revenue  \
0                              Avengers: Endgame  299534  2797800564   
1                                         Avatar   19995  2787965087   
2                   Star Wars: The Force Awakens  140607  2068223624   
3                         Avengers: Infinity War  299536  2046239637   
4                                        Titanic     597  1845034188   
5                                 Jurassic World  135397  1671713208   
6                                  The Lion King  420818  1656943394   
7                                   The Avengers   24428  1519557910   
8                                      Furious 7  168259  1506249360   
9                        Avengers: Age of Ultron   99861  1405403694   
10                                 Black Panther  284054  1346739107   
11  Harry Potter and the Deathly Hallows: Part 2   12445  1341511219   
12                      Star Wars: The Last Jedi  181808  1332539889   
13                                     Frozen II  330457  1330764959   
14                Jurassic World: Fallen Kingdom  351286  1303459585   
15                                        Frozen  109445  1274219009   
16                          Beauty and the Beast  321612  1263521126   
17                                 Incredibles 2  260513  1241891456   

                                               genres  \
0   [{'id': 12, 'name': 'Adventure'}, {'id': 878, ...   
1   [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
2   [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
3   [{'id': 12, 'name': 'Adventure'}, {'id': 28, '...   
4   [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...   
5   [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
6   [{'id': 12, 'name': 'Adventure'}, {'id': 10751...   
7   [{'id': 878, 'name': 'Science Fiction'}, {'id'...   
8   [{'id': 28, 'name': 'Action'}, {'id': 53, 'nam...   
9   [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
10  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
11  [{'id': 14, 'name': 'Fantasy'}, {'id': 12, 'na...   
12  [{'id': 14, 'name': 'Fantasy'}, {'id': 12, 'na...   
13  [{'id': 16, 'name': 'Animation'}, {'id': 10751...   
14  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
15  [{'id': 16, 'name': 'Animation'}, {'id': 12, '...   
16  [{'id': 10751, 'name': 'Family'}, {'id': 14, '...   
17  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   

                                belongs_to_collection  runtime  
0   {'id': 86311, 'name': 'The Avengers Collection...      181  
1   {'id': 87096, 'name': 'Avatar Collection', 'po...      162  
2   {'id': 10, 'name': 'Star Wars Collection', 'po...      136  
3   {'id': 86311, 'name': 'The Avengers Collection...      149  
4                                                None      194  
5   {'id': 328, 'name': 'Jurassic Park Collection'...      124  
6                                                None      118  
7   {'id': 86311, 'name': 'The Avengers Collection...      143  
8   {'id': 9485, 'name': 'The Fast and the Furious...      137  
9   {'id': 86311, 'name': 'The Avengers Collection...      141  
10  {'id': 529892, 'name': 'Black Panther Collecti...      134  
11  {'id': 1241, 'name': 'Harry Potter Collection'...      130  
12  {'id': 10, 'name': 'Star Wars Collection', 'po...      152  
13  {'id': 386382, 'name': 'Frozen Collection', 'p...      104  
14  {'id': 328, 'name': 'Jurassic Park Collection'...      129  
15  {'id': 386382, 'name': 'Frozen Collection', 'p...      102  
16                                               None      129  
17  {'id': 468222, 'name': 'The Incredibles Collec...      118

## Working with APIs and JSON (Part 1)

In [43]:
api_key = "api_key=6dfd66d118d0bca148742d8e0b8489a1" 

In [44]:
import pandas as pd
import requests
pd.options.display.max_columns = 30

2. __Create an account__ on https://www.themoviedb.org/

3. Get your personal __API Key__

4. __API-Request__ (movie module): Load all available information for the movie with __movie id = 140607__ into a Pandas DataFrame. <br> See https://developers.themoviedb.org/3/movies/get-movie-details for more information

In [45]:
movie_id = 157336

In [46]:
movie_api = "https://api.themoviedb.org/3/movie/{}?"
movie_api

'https://api.themoviedb.org/3/movie/{}?'

In [47]:
url = movie_api.format(movie_id) + api_key
url

'https://api.themoviedb.org/3/movie/157336?api_key=6dfd66d118d0bca148742d8e0b8489a1'

In [48]:
r = requests.get(url)
r

<Response [200]>

In [49]:
data = r.json()

In [50]:
data

{'adult': False,
 'backdrop_path': '/ln2Gre4IYRhpjuGVybbtaF4CLo5.jpg',
 'belongs_to_collection': None,
 'budget': 165000000,
 'genres': [{'id': 12, 'name': 'Adventure'},
  {'id': 18, 'name': 'Drama'},
  {'id': 878, 'name': 'Science Fiction'}],
 'homepage': 'http://www.interstellarmovie.net/',
 'id': 157336,
 'imdb_id': 'tt0816692',
 'original_language': 'en',
 'original_title': 'Interstellar',
 'overview': 'The adventures of a group of explorers who make use of a newly discovered wormhole to surpass the limitations on human space travel and conquer the vast distances involved in an interstellar voyage.',
 'popularity': 139.727,
 'poster_path': '/gEU2QniE6E77NI6lCU6MxlNBvIx.jpg',
 'production_companies': [{'id': 923,
   'logo_path': '/8M99Dkt23MjQMTTWukq4m5XsEuo.png',
   'name': 'Legendary Pictures',
   'origin_country': 'US'},
  {'id': 9996,
   'logo_path': '/3tvBqYsBhxWeHlu62SIJ1el93O7.png',
   'name': 'Syncopy',
   'origin_country': 'GB'},
  {'id': 13769,
   'logo_path': None,
   'na

In [51]:
type(data)

dict

In [52]:
pd.DataFrame(data)

ValueError: All arrays must be of the same length

In [53]:
pd.Series(data)

adult                                                                False
backdrop_path                             /ln2Gre4IYRhpjuGVybbtaF4CLo5.jpg
belongs_to_collection                                                 None
budget                                                           165000000
genres                   [{'id': 12, 'name': 'Adventure'}, {'id': 18, '...
homepage                                 http://www.interstellarmovie.net/
id                                                                  157336
imdb_id                                                          tt0816692
original_language                                                       en
original_title                                                Interstellar
overview                 The adventures of a group of explorers who mak...
popularity                                                         139.727
poster_path                               /gEU2QniE6E77NI6lCU6MxlNBvIx.jpg
production_companies     

In [55]:
df = pd.Series(data).to_frame().T
df

adult                     backdrop_path belongs_to_collection     budget  \
0  False  /ln2Gre4IYRhpjuGVybbtaF4CLo5.jpg                  None  165000000   

                                              genres  \
0  [{'id': 12, 'name': 'Adventure'}, {'id': 18, '...   

                            homepage      id    imdb_id original_language  \
0  http://www.interstellarmovie.net/  157336  tt0816692                en   

  original_title                                           overview  \
0   Interstellar  The adventures of a group of explorers who mak...   

  popularity                       poster_path  \
0    139.727  /gEU2QniE6E77NI6lCU6MxlNBvIx.jpg   

                                production_companies  \
0  [{'id': 923, 'logo_path': '/8M99Dkt23MjQMTTWuk...   

                                production_countries release_date    revenue  \
0  [{'iso_3166_1': 'GB', 'name': 'United Kingdom'...   2014-11-05  701729206   

  runtime                                   spoken_languages    status  \
0     169  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   

                                             tagline         title  video  \
0  Mankind was born on Earth. It was never meant ...  Interstellar  False   

  vote_average vote_count  
0        8.429      33789

In [56]:
pd.json_normalize(data, sep = "_")

adult                     backdrop_path belongs_to_collection     budget  \
0  False  /ln2Gre4IYRhpjuGVybbtaF4CLo5.jpg                  None  165000000   

                                              genres  \
0  [{'id': 12, 'name': 'Adventure'}, {'id': 18, '...   

                            homepage      id    imdb_id original_language  \
0  http://www.interstellarmovie.net/  157336  tt0816692                en   

  original_title                                           overview  \
0   Interstellar  The adventures of a group of explorers who mak...   

   popularity                       poster_path  \
0     139.727  /gEU2QniE6E77NI6lCU6MxlNBvIx.jpg   

                                production_companies  \
0  [{'id': 923, 'logo_path': '/8M99Dkt23MjQMTTWuk...   

                                production_countries release_date    revenue  \
0  [{'iso_3166_1': 'GB', 'name': 'United Kingdom'...   2014-11-05  701729206   

   runtime                                   spoken_languages    status  \
0      169  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   

                                             tagline         title  video  \
0  Mankind was born on Earth. It was never meant ...  Interstellar  False   

   vote_average  vote_count  
0         8.429       33789

In [57]:
pd.json_normalize(data = data, record_path =  "genres", meta = "title")

id             name         title
0   12        Adventure  Interstellar
1   18            Drama  Interstellar
2  878  Science Fiction  Interstellar

In [58]:
pd.json_normalize(data = data, record_path =  "production_companies", meta = "title")

id                         logo_path                    name  \
0    923  /8M99Dkt23MjQMTTWukq4m5XsEuo.png      Legendary Pictures   
1   9996  /3tvBqYsBhxWeHlu62SIJ1el93O7.png                 Syncopy   
2  13769                              None  Lynda Obst Productions   

  origin_country         title  
0             US  Interstellar  
1             GB  Interstellar  
2                 Interstellar

## Working with APIs and JSON (Part 2)

5. __API-Request__ (discover module): Load all movies with __release date between 2020-01-01 and 2020-02-29__ into a Pandas DataFrame. <br>
See https://www.themoviedb.org/documentation/api/discover and https://developers.themoviedb.org/3/discover/movie-discover for more information.

In [59]:
discover_api = "https://api.themoviedb.org/3/discover/movie?"

##  Importing and Saving the Movies Dataset (Best Practice)

In [68]:
query = "&primary_release_date.gte=2020-01-01&primary_release_date.lte=2020-02-29&page=2"

In [69]:
url = discover_api+api_key+query

In [70]:
data = requests.get(url).json()

In [71]:
data

{'page': 2,
 'results': [{'adult': False,
   'backdrop_path': '/f6YSikiGgSx2IEHiDAWwTt3uEki.jpg',
   'genre_ids': [35, 18, 10749],
   'id': 802119,
   'original_language': 'en',
   'original_title': 'All About Sex',
   'overview': 'Three friends in their mid-20s struggle to navigate their professional and personal lives, colliding head on with the messy, hilarious and dreadful growing pangs of adulthood.',
   'popularity': 21.168,
   'poster_path': '/cNTwgukBFEYtYNMGsKxgrCFm1xF.jpg',
   'release_date': '2020-01-16',
   'title': 'All About Sex',
   'video': False,
   'vote_average': 6.2,
   'vote_count': 31},
  {'adult': False,
   'backdrop_path': '/27nn2YJBwY7a1etZ0AiIcOrIgu3.jpg',
   'genre_ids': [10749, 18],
   'id': 664413,
   'original_language': 'pl',
   'original_title': '365 dni',
   'overview': 'A woman falls victim to a dominant mafia boss, who imprisons her and gives her one year to fall in love with him.',
   'popularity': 22.741,
   'poster_path': '/6KwrHucIE3CvNT7kTm2MAlZ4

In [72]:
pd.DataFrame(data)

page                                            results  total_pages  \
0      2  {'adult': False, 'backdrop_path': '/f6YSikiGgS...          289   
1      2  {'adult': False, 'backdrop_path': '/27nn2YJBwY...          289   
2      2  {'adult': False, 'backdrop_path': '/42gqGeyXh4...          289   
3      2  {'adult': False, 'backdrop_path': '/dxdN3RxJRV...          289   
4      2  {'adult': False, 'backdrop_path': '/lOEMTu6Qhw...          289   
5      2  {'adult': False, 'backdrop_path': '/tQprwIKzFx...          289   
6      2  {'adult': False, 'backdrop_path': '/pQdZmXq0vv...          289   
7      2  {'adult': False, 'backdrop_path': '/xYiPGeQd4A...          289   
8      2  {'adult': False, 'backdrop_path': '/tcrNJfyNEI...          289   
9      2  {'adult': False, 'backdrop_path': '/bpHseZ8Vwh...          289   
10     2  {'adult': False, 'backdrop_path': '/jRweromdBo...          289   
11     2  {'adult': False, 'backdrop_path': '/wEVSupeNO7...          289   
12     2  {'adult': False, 'backdrop_path': '/6mKAKhj8PO...          289   
13     2  {'adult': False, 'backdrop_path': '/lWJJ5GDUFr...          289   
14     2  {'adult': False, 'backdrop_path': '/bFxGD1TKrJ...          289   
15     2  {'adult': False, 'backdrop_path': '/kLsrE6q4ld...          289   
16     2  {'adult': False, 'backdrop_path': '/rh0Q3iM19n...          289   
17     2  {'adult': False, 'backdrop_path': '/dUasP5TGFk...          289   
18     2  {'adult': False, 'backdrop_path': '/77xFxLtD2j...          289   
19     2  {'adult': False, 'backdrop_path': '/2EOy2YSJbz...          289   

    total_results  
0            5777  
1            5777  
2            5777  
3            5777  
4            5777  
5            5777  
6            5777  
7            5777  
8            5777  
9            5777  
10           5777  
11           5777  
12           5777  
13           5777  
14           5777  
15           5777  
16           5777  
17           5777  
18           5777  
19           5777

In [73]:
pd.DataFrame(data["results"])

adult                     backdrop_path           genre_ids      id  \
0   False  /f6YSikiGgSx2IEHiDAWwTt3uEki.jpg     [35, 18, 10749]  802119   
1   False  /27nn2YJBwY7a1etZ0AiIcOrIgu3.jpg         [10749, 18]  664413   
2   False  /42gqGeyXh4caHP19Tqf8DMRgc7r.jpg         [35, 10749]  556678   
3   False  /dxdN3RxJRVYzFxVSR0fNhSBqBSL.jpg                [99]  682666   
4   False  /lOEMTu6QhwrTvQdV8kuVgMOABth.jpg                [18]  880956   
5   False   /tQprwIKzFx83JpM5dVrGRC4UpF.jpg  [16, 12, 14, 9648]  573730   
6   False  /pQdZmXq0vvKP1RYUrDHQkzMlS4x.jpg    [16, 28, 12, 14]  571265   
7   False  /xYiPGeQd4AVuLxR9cGdce0B4w3y.jpg        [53, 18, 27]  715756   
8   False  /tcrNJfyNEIqaBR8Ogkgnq5xQJnf.jpg         [10749, 18]  342470   
9   False  /bpHseZ8Vwh3nk7cowKxQ68Dm2Xq.jpg        [28, 18, 53]  616803   
10  False  /jRweromdBoFchxYF6vn6E4wEJla.jpg         [10749, 35]  565426   
11  False  /wEVSupeNO7EjjCFieEv2zt4oJei.jpg        [35, 53, 80]  609242   
12  False  /6mKAKhj8POVGqV1GsroS5mGIUe9.jpg        [14, 28, 12]  666750   
13  False  /lWJJ5GDUFrOXhhOICRDBNO4eGxV.jpg            [27, 53]  571625   
14  False  /bFxGD1TKrJq3Dr5Pjnn739gBhQF.jpg     [10751, 28, 35]  592834   
15  False  /kLsrE6q4ld144ECBkydZ4dV8t9T.jpg          [27, 9648]  648990   
16  False  /rh0Q3iM19n0UxIuAB8W3wXKO5O9.jpg        [28, 12, 53]  531299   
17  False  /dUasP5TGFk1dhTgtNhk5uLRPRx6.jpg     [10752, 36, 18]  592279   
18  False  /77xFxLtD2jGpOrcjJbCez2wZWX1.jpg             [10749]  698818   
19  False  /2EOy2YSJbzKvlzKpEEKULVYZGlN.jpg         [10749, 35]  658760   

   original_language                          original_title  \
0                 en                           All About Sex   
1                 pl                                 365 dni   
2                 en                                   Emma.   
3                 en                     Back to the Titanic   
4                 ja                              美乳夜曲 乱れる白肌   
5                 ja                      劇場版メイドインアビス 深き魂の黎明   
6                 ja            デジモンアドベンチャー LAST EVOLUTION 絆   
7                 ja                                    性の劇薬   
8                 en                   All the Bright Places   
9                 sv                        Breaking Surface   
10                en  To All the Boys: P.S. I Still Love You   
11                es                       El robo del siglo   
12                en                  Dragonheart: Vengeance   
13                ko                                     클로젯   
14                en                                  My Spy   
15                es                                 La casa   
16                en                              Kill Chain   
17                ru                              Калашников   
18                ko                         음란 주택: 노예가 된 아내   
19                de                              Isi & Ossi   

                                             overview  popularity  \
0   Three friends in their mid-20s struggle to nav...      21.168   
1   A woman falls victim to a dominant mafia boss,...      22.741   
2   In 1800s England, a well-meaning but selfish y...      32.029   
3   Back to the Titanic documents the first manned...      20.316   
4   At night, Luna after taking a bath is humming ...      17.911   
5   A continuation of the epic adventure of plucky...      26.655   
6   Tai is now a university student, living alone,...      24.107   
7   Katsuragi, an elite office worker who fell fro...      23.187   
8   Two teens facing personal struggles form a pow...      24.295   
9   Two Swedish/Norwegian half sisters go on a win...      22.774   
10  Lara Jean and Peter have just taken their roma...      27.506   
11  In 2006, a group of thieves performed what is ...      26.707   
12  Lukas, a young farmer whose family is killed b...      27.154   
13  After moving into a new house, a young girl be...      17.582   
14  A hardened CIA operative finds himself at th

6. __API-Request__ (movie module): Load all available information for the movies with movie id = [__299534, 19995, 140607, 299536, 597, 135397, 420818, 24428, 168259, 99861, 284054, 12445, 181808, 330457, 351286, 109445, 321612, 260513__] into a Pandas DataFrame and __save the dataset in a local json file__.

# +++++++++ See some Hints below +++++++++++++

# ++++++++++++++ Hints +++++++++++++++++++++

__Hints for 1.__ <br>
To load json files you can use 

In [ ]:
with open("filename.json") as f:
    data = json.load(f)

and 

In [ ]:
pd.DataFrame(data), pd.read_json(filename.json), pd.json_normalize(data)

the json files have the following orientation (important when using pd.read_json()):
- blockbusters.json -> record
- blockbusters2.json -> column
- blockbusters3.json -> split 

__Hints for 4., 5., 6.__<br>
Make API GET-requests with the library requests (import requests):

In [ ]:
data = requests.get(url).json()

__Hints for 4. and 6.,__ <br> url structure for movie module:

"https://api.themoviedb.org/3/movie/insert_movie_id?api_key=insert_api_key" (replace "insert_movie_id" with movie id and "insert_api_key" with your personal api-key)

__Hints for 5.__<br>
url structure for discover module:

"https://api.themoviedb.org/3/discover/movie?api_key=insert_api_key&query1&query2..." (replace "insert_api_key" with your personal api-key and add appropriate queries)